source/simulator/parameters.cc

```diff
@@ -1961,11 +1964,18 @@ namespace aspect
 
       if (std::find(compositional_field_methods.begin(), compositional_field_methods.end(), AdvectionFieldMethod::fem_darcy_field)
           != compositional_field_methods.end())
-        AssertThrow (std::find(names_of_compositional_fields.begin(),
-                               names_of_compositional_fields.end(),
-                               "porosity")
+        AssertThrow (std::find(names_of_compositional_fields.begin(), names_of_compositional_fields.end(), "porosity")
                      != names_of_compositional_fields.end(),
                      ExcMessage ("The Darcy advection field method only works if there is a compositional field named 'porosity'"));
+      if ((std::find(compositional_field_methods.begin(), compositional_field_methods.end(), AdvectionFieldMethod::fem_darcy_field)
+           != compositional_field_methods.end()) && (std::find(names_of_compositional_fields.begin(), names_of_compositional_fields.end(), "porosity") != names_of_compositional_fields.end()) )
+        for (unsigned int i=0; i<n_compositional_fields; ++i)
+
+          {
+            AssertThrow ((((compositional_field_methods[i] != AdvectionFieldMethod::fem_darcy_field) &&
+                           (names_of_compositional_fields[i] == "porosity")) == false),
+                         ExcMessage ("When using the darcy field method, the porosity field must be advected with the darcy method."));
+          }
```

Comment:
I think you can make this simpler and shorter:
```suggestion
        {
          const unsigned int porosity_idx = std::find(names_of_compositional_fields.begin(), names_of_compositional_fields.end(), "porosity");
          AssertThrow (porosity_idx != names_of_compositional_fields.end(),
                       ExcMessage ("The Darcy advection field method only works if there is a compositional field named 'porosity'"));
          AssertThrow (compositional_field_methods[porosity_idx] == AdvectionFieldMethod::fem_darcy_field,
                       ExcMessage ("When using the darcy field method, the porosity field must be advected with the darcy method."));
        }
```

Reply:
I ran into an issue when just trying to implement what you have here, it might still work but std::find returns an iterator and not an unsigned int, I was not sure how best to get around this without making the code look messier again.